File: python/bark_sdk.py
import requests
import json
import urllib.parse
from typing import Optional, Dict, Any, Union


class BarkException(Exception):
    """Exception raised for Bark SDK errors."""
    
    def __init__(self, message: str, status_code: Optional[int] = None, response: Optional[Dict[str, Any]] = None):
        self.message = message
        self.status_code = status_code
        self.response = response
        super().__init__(self.message)
    
    def __str__(self) -> str:
        if self.status_code:
            return f"{self.message} (Status code: {self.status_code})"
        return self.message


class BarkNetworkException(BarkException):
    """Exception raised for network errors when communicating with Bark server."""
    pass


class BarkServerException(BarkException):
    """Exception raised for server errors returned by the Bark server."""
    pass


class BarkClientException(BarkException):
    """Exception raised for client-side validation errors."""
    pass


class BarkNotification:
    """
    A Python SDK for Bark iOS push notification service.
    Bark allows sending custom notifications to your iPhone.
    
    Official GitHub: https://github.com/Finb/Bark
    """

    DEFAULT_SERVER = "https://api.day.app"
    
    def __init__(self, key: str, server_url: Optional[str] = None):
        """
        Initialize the Bark notification client.
        
        Args:
            key (str): Your Bark key from the Bark iOS app
            server_url (str, optional): Custom server URL if you're self-hosting Bark
        """
        if not key:
            raise BarkClientException("Bark key cannot be empty")
            
        self.key = key
        self.server_url = server_url or self.DEFAULT_SERVER
        self.base_url = f"{self.server_url}/{self.key}"
    
    def send(self, 
             body: str, 
             title: Optional[str] = None, 
             subtitle: Optional[str] = None, 
             url: Optional[str] = None, 
             group: Optional[str] = None, 
             icon: Optional[str] = None, 
             sound: Optional[str] = None, 
             call: Optional[bool] = None, 
             level: Optional[str] = None, 
             is_archive: Optional[bool] = None, 
             copy: Optional[str] = None, 
             ciphertext: Optional[str] = None) -> Dict[str, Any]:
        """
        Send a notification to your device.
        
        Args:
            body (str): Main notification content
            title (str, optional): Notification title
            subtitle (str, optional): Notification subtitle
            url (str, optional): URL to open when notification is tapped
            group (str, optional): Group identifier for notifications
            icon (str, optional): Custom icon URL (iOS 15+ only)
            sound (str, optional): Custom notification sound
            call (bool, optional): If True, plays sound repeatedly for 30 seconds
            level (str, optional): Notification importance level: "active", "timeSensitive", "passive", or "critical"
            is_archive (bool, optional): Whether to archive the notification
            copy (str, optional): Text to copy to clipboard when notification is pressed
            ciphertext (str, optional): Encrypted notification content
            
        Returns:
            dict: Response from the Bark server
            
        Raises:
            BarkClientException: If validation fails or parameters are invalid
            BarkNetworkException: If there's a network error
            BarkServerException: If the server returns an error response
        """
        if not body:
            raise BarkClientException("Notification body cannot be empty")
            
        # Validate level if provided
        if level and level not in ["active", "timeSensitive", "passive", "critical"]:
            raise BarkClientException(
                "Invalid level value. Must be one of: active, timeSensitive, passive, critical"
            )
        
        try:
            # Build endpoint based on provided parameters
            endpoint = self._build_endpoint(body, title, subtitle)
            
            # Prepare parameters
            params = {}
            if url:
                params['url'] = url
            if group:
                params['group'] = group
            if icon:
                params['icon'] = icon
            if sound:
                params['sound'] = sound
            if call:
                params['call'] = 1 if call else 0
            if level:
                params['level'] = level
            if is_archive is not None:
                params['isArchive'] = 1 if is_archive else 0
            if copy:
                params['copy'] = copy
            if ciphertext:
                params['ciphertext'] = ciphertext
                
            # Send the request
            try:
                response = requests.get(endpoint, params=params, timeout=10)
            except requests.RequestException as e:
                raise BarkNetworkException(f"Network error: {str(e)}")
            
            # Parse response
            return self._handle_response(response)
            
        except (ValueError, TypeError) as e:
            raise BarkClientException(f"Invalid parameter: {str(e)}")
    
    def send_post(self, 
                 body: str, 
                 title: Optional[str] = None, 
                 subtitle: Optional[str] = None, 
                 **kwargs) -> Dict[str, Any]:
        """
        Send a notification using POST request instead of GET.
        
        Args:
            body (str): Main notification content
            title (str, optional): Notification title
            subtitle (str, optional): Notification subtitle
            **kwargs: Additional parameters (see send method)
            
        Returns:
            dict: Response from the Bark server
            
        Raises:
            BarkClientException: If validation fails or parameters are invalid
            BarkNetworkException: If there's a network error
            BarkServerException: If the server returns an error response
        """
        if not body:
            raise BarkClientException("Notification body cannot be empty")
            
        # Validate level if provided
        if 'level' in kwargs and kwargs['level'] not in ["active", "timeSensitive", "passive", "critical"]:
            raise BarkClientException(
                "Invalid level value. Must be one of: active, timeSensitive, passive, critical"
            )
        
        try:
            # Prepare data
            data = {
                "body": body
            }
            
            if title:
                data["title"] = title
            if subtitle:
                data["subtitle"] = subtitle
                
            # Add other parameters
            for key, value in kwargs.items():
                data[key] = value
                
            # Send the request
            try:
                response = requests.post(f"{self.base_url}", json=data, timeout=10)
            except requests.RequestException as e:
                raise BarkNetworkException(f"Network error: {str(e)}")
            
            # Parse response
            return self._handle_response(response)
            
        except (ValueError, TypeError) as e:
            raise BarkClientException(f"Invalid parameter: {str(e)}")
    
    def _build_endpoint(self, body: str, title: Optional[str] = None, subtitle: Optional[str] = None) -> str:
        """
        Build the endpoint URL based on provided parameters.
        
        Args:
            body (str): Main notification content
            title (str, optional): Notification title
            subtitle (str, optional): Notification subtitle
            
        Returns:
            str: The constructed endpoint URL
        """
        try:
            if title and subtitle:
                return f"{self.base_url}/{urllib.parse.quote(title)}/{urllib.parse.quote(subtitle)}/{urllib.parse.quote(body)}"
            elif title:
                return f"{self.base_url}/{urllib.parse.quote(title)}/{urllib.parse.quote(body)}"
            else:
                return f"{self.base_url}/{urllib.parse.quote(body)}"
        except Exception as e:
            raise BarkClientException(f"Error building endpoint URL: {str(e)}")
    
    def _handle_response(self, response: requests.Response) -> Dict[str, Any]:
        """
        Handle the response from the Bark server.
        
        Args:
            response: Response from the Bark server
            
        Returns:
            dict: Parsed response
            
        Raises:
            BarkServerException: If the server returns an error response
        """
        try:
            # Check HTTP status code
            if response.status_code != 200:
                raise BarkServerException(
                    f"Server returned error: {response.text}",
                    status_code=response.status_code
                )
            
            # Parse JSON response
            json_response = response.json()
            
            # Check API response code
            if json_response.get('code') != 200:
                raise BarkServerException(
                    f"API error: {json_response.get('message', 'Unknown error')}",
                    status_code=response.status_code,
                    response=json_response
                )
                
            return json_response
            
        except json.JSONDecodeError:
            raise BarkServerException(
                f"Invalid JSON response: {response.text}",
                status_code=response.status_code
            )


# Usage example
if __name__ == "__main__":
    try:
        # Create a notification client with your Bark key
        bark = BarkNotification(key="YOUR_BARK_KEY")
        
        # Send a simple notification
        response = bark.send(body="Hello from Python SDK!")
        print(response)
        
        # Send a notification with more parameters
        response = bark.send(
            title="Notification Title",
            subtitle="Notification Subtitle",
            body="This is the main notification content.\nWith multiple lines!",
            url="https://example.com",
            group="example-notifications",
            sound="alarm",
            level="timeSensitive"
        )
        print(response)
    except BarkClientException as e:
        print(f"Client error: {e}")
    except BarkNetworkException as e:
        print(f"Network error: {e}")
    except BarkServerException as e:
        print(f"Server error: {e}")
        if e.response:
            print(f"Response details: {e.response}")
    except Exception as e:
        print(f"Unexpected error: {e}") 

File: javascript/bark_sdk.js
/**
 * Bark Notification JavaScript SDK
 * 
 * A JavaScript client for the Bark iOS push notification service.
 * Send custom notifications to your iPhone easily.
 * 
 * Official GitHub: https://github.com/Finb/Bark
 */

/**
 * Base class for Bark SDK exceptions
 */
class BarkException extends Error {
  /**
   * Create a new BarkException
   * @param {string} message - Error message
   * @param {number} [statusCode] - HTTP status code
   * @param {object} [response] - Response object
   */
  constructor(message, statusCode = null, response = null) {
    super(message);
    this.name = this.constructor.name;
    this.statusCode = statusCode;
    this.response = response;
    
    // This is for backward compatibility with native Error objects
    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, this.constructor);
    }
  }
}

/**
 * Exception for network errors when communicating with Bark server
 */
class BarkNetworkException extends BarkException {
  constructor(message, statusCode = null, response = null) {
    super(message, statusCode, response);
  }
}

/**
 * Exception for server errors returned by the Bark server
 */
class BarkServerException extends BarkException {
  constructor(message, statusCode = null, response = null) {
    super(message, statusCode, response);
  }
}

/**
 * Exception for client-side validation errors
 */
class BarkClientException extends BarkException {
  constructor(message) {
    super(message);
  }
}

class BarkNotification {
  /**
   * Create a new Bark notification client
   * @param {string} key - Your Bark key from the Bark iOS app
   * @param {string} [serverUrl="https://api.day.app"] - Custom server URL if you're self-hosting Bark
   * @throws {BarkClientException} If key is empty or invalid
   */
  constructor(key, serverUrl = 'https://api.day.app') {
    if (!key) {
      throw new BarkClientException('Bark key cannot be empty');
    }
    
    this.key = key;
    this.serverUrl = serverUrl;
    this.baseUrl = `${this.serverUrl}/${this.key}`;
  }

  /**
   * Send a notification using GET request
   * 
   * @param {Object} options - Notification options
   * @param {string} options.body - Main notification content
   * @param {string} [options.title] - Notification title
   * @param {string} [options.subtitle] - Notification subtitle
   * @param {string} [options.url] - URL to open when notification is tapped
   * @param {string} [options.group] - Group identifier for notifications
   * @param {string} [options.icon] - Custom icon URL (iOS 15+ only)
   * @param {string} [options.sound] - Custom notification sound
   * @param {boolean} [options.call] - If true, plays sound repeatedly for 30 seconds
   * @param {string} [options.level] - Notification importance level: "active", "timeSensitive", "passive", or "critical"
   * @param {boolean} [options.isArchive] - Whether to archive the notification
   * @param {string} [options.copy] - Text to copy to clipboard when notification is pressed
   * @param {string} [options.ciphertext] - Encrypted notification content
   * @returns {Promise<Object>} - Response from the Bark server
   * @throws {BarkClientException} If validation fails or parameters are invalid
   * @throws {BarkNetworkException} If there's a network error
   * @throws {BarkServerException} If the server returns an error response
   */
  async send({
    body,
    title,
    subtitle,
    url,
    group,
    icon,
    sound,
    call,
    level,
    isArchive,
    copy,
    ciphertext
  }) {
    if (!body) {
      throw new BarkClientException('Notification body is required');
    }
    
    // Validate level if provided
    if (level && !['active', 'timeSensitive', 'passive', 'critical'].includes(level)) {
      throw new BarkClientException('Invalid level value. Must be one of: active, timeSensitive, passive, critical');
    }

    try {
      // Build endpoint based on provided parameters
      const endpoint = this._buildEndpoint(body, title, subtitle);
      
      // Prepare parameters
      const params = new URLSearchParams();
      if (url) params.append('url', url);
      if (group) params.append('group', group);
      if (icon) params.append('icon', icon);
      if (sound) params.append('sound', sound);
      if (call) params.append('call', call ? '1' : '0');
      if (level) params.append('level', level);
      if (isArchive !== undefined) params.append('isArchive', isArchive ? '1' : '0');
      if (copy) params.append('copy', copy);
      if (ciphertext) params.append('ciphertext', ciphertext);
      
      // Build the final URL
      const queryString = params.toString();
      const requestUrl = queryString ? `${endpoint}?${queryString}` : endpoint;
      
      // Send the request
      const response = await this._fetchWithTimeout(requestUrl);
      
      // Handle the response
      return this._handleResponse(response);
    } catch (error) {
      // Rethrow if it's already one of our custom exceptions
      if (error instanceof BarkException) {
        throw error;
      }
      
      // Otherwise wrap it in a BarkNetworkException
      throw new BarkNetworkException(`Request failed: ${error.message}`);
    }
  }

  /**
   * Send a notification using POST request
   * 
   * @param {Object} options - Notification options (see send method for details)
   * @returns {Promise<Object>} - Response from the Bark server
   * @throws {BarkClientException} If validation fails or parameters are invalid
   * @throws {BarkNetworkException} If there's a network error
   * @throws {BarkServerException} If the server returns an error response
   */
  async sendPost(options) {
    if (!options.body) {
      throw new BarkClientException('Notification body is required');
    }
    
    // Validate level if provided
    if (options.level && !['active', 'timeSensitive', 'passive', 'critical'].includes(options.level)) {
      throw new BarkClientException('Invalid level value. Must be one of: active, timeSensitive, passive, critical');
    }
    
    try {
      // Send the request
      const response = await this._fetchWithTimeout(this.baseUrl, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json'
        },
        body: JSON.stringify(options)
      });
      
      // Handle the response
      return this._handleResponse(response);
    } catch (error) {
      // Rethrow if it's already one of our custom exceptions
      if (error instanceof BarkException) {
        throw error;
      }
      
      // Otherwise wrap it in a BarkNetworkException
      throw new BarkNetworkException(`Request failed: ${error.message}`);
    }
  }

  /**
   * Fetch with timeout to prevent hanging requests
   * 
   * @private
   * @param {string} url - The URL to fetch
   * @param {Object} [options] - Fetch options
   * @param {number} [timeout=10000] - Timeout in milliseconds
   * @returns {Promise<Response>} - Fetch response
   * @throws {BarkNetworkException} If the request times out or fails
   */
  async _fetchWithTimeout(url, options = {}, timeout = 10000) {
    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), timeout);
    
    try {
      const response = await fetch(url, {
        ...options,
        signal: controller.signal
      });
      
      return response;
    } catch (error) {
      if (error.name === 'AbortError') {
        throw new BarkNetworkException(`Request timed out after ${timeout}ms`);
      }
      throw new BarkNetworkException(`Network error: ${error.message}`);
    } finally {
      clearTimeout(timeoutId);
    }
  }

  /**
   * Handle the response from the Bark server
   * 
   * @private
   * @param {Response} response - Fetch response
   * @returns {Promise<Object>} - Parsed response
   * @throws {BarkServerException} If the server returns an error response
   */
  async _handleResponse(response) {
    // Check HTTP status code
    if (!response.ok) {
      let errorText;
      try {
        errorText = await response.text();
      } catch (e) {
        errorText = 'Unable to read error response';
      }
      
      throw new BarkServerException(
        `Server returned error: ${errorText}`,
        response.status
      );
    }
    
    // Parse JSON response
    try {
      const jsonResponse = await response.json();
      
      // Check API response code
      if (jsonResponse.code !== 200) {
        throw new BarkServerException(
          `API error: ${jsonResponse.message || 'Unknown error'}`,
          response.status,
          jsonResponse
        );
      }
      
      return jsonResponse;
    } catch (error) {
      if (error instanceof BarkException) {
        throw error;
      }
      
      throw new BarkServerException(
        `Invalid JSON response: ${error.message}`,
        response.status
      );
    }
  }

  /**
   * Build the endpoint URL based on provided parameters
   * 
   * @private
   * @param {string} body - Main notification content
   * @param {string} [title] - Notification title
   * @param {string} [subtitle] - Notification subtitle
   * @returns {string} - The constructed endpoint URL
   * @throws {BarkClientException} If there's an error building the URL
   */
  _buildEndpoint(body, title, subtitle) {
    try {
      if (title && subtitle) {
        return `${this.baseUrl}/${encodeURIComponent(title)}/${encodeURIComponent(subtitle)}/${encodeURIComponent(body)}`;
      } else if (title) {
        return `${this.baseUrl}/${encodeURIComponent(title)}/${encodeURIComponent(body)}`;
      } else {
        return `${this.baseUrl}/${encodeURIComponent(body)}`;
      }
    } catch (error) {
      throw new BarkClientException(`Error building endpoint URL: ${error.message}`);
    }
  }
}

// Export the main class and exception classes
if (typeof module !== 'undefined' && module.exports) {
  module.exports = BarkNotification;
  module.exports.BarkException = BarkException;
  module.exports.BarkNetworkException = BarkNetworkException;
  module.exports.BarkServerException = BarkServerException;
  module.exports.BarkClientException = BarkClientException;
}

// Usage example in browser:
/*
const bark = new BarkNotification('YOUR_BARK_KEY');

// Send a simple notification
bark.send({ body: 'Hello from JavaScript SDK!' })
  .then(response => console.log('Notification sent:', response))
  .catch(error => {
    if (error instanceof BarkClientException) {
      console.error('Client error:', error.message);
    } else if (error instanceof BarkNetworkException) {
      console.error('Network error:', error.message);
    } else if (error instanceof BarkServerException) {
      console.error('Server error:', error.message, 'Status:', error.statusCode);
      if (error.response) {
        console.error('Response details:', error.response);
      }
    } else {
      console.error('Unexpected error:', error);
    }
  });
*/ 